### ISSUE TITLE
Implement multiple ensemble style syntax used for datasets in fx variable descriptions for pre processor

### ISSUE THREAD

**Issue by thomascrocker** [ESMValGroup/ESMValCore#1081]
**Is your feature request related to a problem? Please describe.**
I'm writing recipes using a large number of the CORDEX EUR-11 models, and need to do some land sea masking. Unfortunately, for some (not all) of the CORDEX models there are not separate sftlf fx files for each ensemble member, and instead, the fx fields exist under ensemble r0i0p0 whereas the rest of the variables are under their own ensemble numbers r1i1p1 etc. See for example: https://esgf-index1.ceda.ac.uk/esg-search/search/?offset=0&limit=10&type=Dataset&replica=false&latest=true&domain=EUR-11&rcm_name=RCA4&time_frequency=fx%2Cmon&variable=sftlf%2Ctas&project=CORDEX&facets=rcm_name%2Cproject%2Cproduct%2Cdomain%2Cinstitute%2Cdriving_model%2Cexperiment%2Cexperiment_family%2Censemble%2Crcm_version%2Ctime_frequency%2Cvariable%2Cvariable_long_name%2Ccf_standard_name%2Cdata_node&format=application%2Fsolr%2Bjson
This causes the preprocessor to break because it is searching for the fx files under the same ensemble number as the rest of the variables (and they don't exist).
What would be great, would be if the user could use the syntax developed for specifying multiple ensembles in datasets at https://docs.esmvaltool.org/projects/esmvalcore/en/latest/recipe/overview.html#recipe-section-datasets e.g. by specifying something like:
```
mask_landsea:
      mask_out: sea
      fx_variables:
        - { short_name: sftlf, ensemble: "r(0:9)i(0:9)p(0:9)" }
 ```
In the preprocessor. However, the preprocessor interprets the string literally and fails to find the files.
It's not possible to use the Natural Earth dataset with this data, since the data is on a rotated pole and thus has multidimensional "latitude" and "longitude" co-ordinates which is not yet supported for use with shapefiles. (That's another issue I need to look into soon). 

**Would you be able to help out?**
Would you have the time and skills to implement the solution yourself?
I can have a look, but I'll need some pointers on where to get started etc.

**Comment by thomascrocker** [comment]
Tagging @zklaus since I know he has been involved with the regional model work.

**Comment by thomascrocker** [comment]
This appears to be applicable to large numbers of the CMIP5 and 6 datasets too.. i.e. the fx variables are stored under a seperate ensemble r0i0p0 from the other variables

**Comment by thomascrocker** [comment]
Ahhh.. Looks like when not specifying the `fx_variables:` key the pre processor knows to go search different ensemble members.. OK, maybe this is actually an issue with the CORDEX CMOR tables and if I tweak those I can get away with not needing the fx_variables key...

**Comment by thomascrocker** [comment]
Hmmm. So I updated the CORDEX FX table to include the missing sftof entry. Now if I just have the pre processor set as 
```
mask_landsea:
      mask_out: sea
```
The recipe still fails to find the fx file stored under r0i0p0 and then fails because the natural earth shapefile can't be used with the rotated grid
```
2021-04-23 16:14:06,577 UTC [44120] DEBUG   esmvalcore._config._config:154 Retrieving CORDEX configuration
2021-04-23 16:14:06,577 UTC [44120] DEBUG   esmvalcore._recipe:63 If not present: adding keys from CMOR table to {'short_name': 'sftlf', 'preprocessor': 'test', 'start_year': 1980, 'end_year': 1999, 'exp': 'historical', 'variable_group': 'sftlf', 'diagnostic': 'stats', 'institute': 'SMHI', 'driver': 'NCC-NorESM1-M', 'dataset': 'RCA4', 'project': 'CORDEX', 'ensemble': 'r1i1p1', 'mip': 'fx', 'rcm_version': 'v1', 'domain': 'EUR-11', 'recipe_dataset_index': 0, 'alias': 'RCA4', 'original_short_name': 'tas', 'standard_name': 'air_temperature', 'long_name': 'Near-Surface Air Temperature', 'units': 'K', 'modeling_realm': ['atmos'], 'frequency': 'mon', 'filename': '/net/home/h02/tcrocker/code/EUCP_WP5_Lines_of_Evidence/esmvaltool/esmvaltool_output/recipe_cordex_test_20210423_161405/preproc/stats/tas/tas_EUR-11_NCC-NorESM1-M_RCA4_historical_r1i1p1_v1_mon_1980-1999.nc'}
2021-04-23 16:14:06,577 UTC [44120] DEBUG   esmvalcore._recipe:383 For fx variable 'sftlf', found table 'fx'
2021-04-23 16:14:06,577 UTC [44120] DEBUG   esmvalcore._config._config:154 Retrieving CORDEX configuration
2021-04-23 16:14:06,577 UTC [44120] DEBUG   esmvalcore._data_finder:102 {'exp', 'latestversion', 'domain', 'ensemble', 'dataset', 'institute', 'driver', 'mip', 'short_name', 'rcm_version'}
2021-04-23 16:14:06,597 UTC [44120] DEBUG   esmvalcore._data_finder:220 Skipping non-existent /project/champ/data/cordex/output/EUR-11/SMHI/NCC-NorESM1-M/historical/r1i1p1/RCA4/v1/fx/sftlf/{latestversion}
2021-04-23 16:14:06,598 UTC [44120] DEBUG   esmvalcore._config._config:154 Retrieving CORDEX configuration
2021-04-23 16:14:06,598 UTC [44120] DEBUG   esmvalcore._data_finder:102 {'exp', 'domain', 'ensemble', 'dataset', 'institute', 'driver', 'mip', 'short_name', 'rcm_version'}
2021-04-23 16:14:06,598 UTC [44120] DEBUG   esmvalcore._data_finder:23 Looking for files matching ['sftlf_EUR-11_NCC-NorESM1-M_historical_r1i1p1_SMHI-RCA4_v1_fx*.nc'] in []
2021-04-23 16:14:06,598 UTC [44120] WARNING esmvalcore._recipe:401 Missing data for fx variable 'sftlf'
```

**Comment by thomascrocker** [comment]
Ahh, OK, I understand what is going on here. There is a fix hardcoded for CMIP5 here
https://github.com/ESMValGroup/ESMValCore/blob/22043381b6405c17c407a35bff8bf6ebc538b07d/esmvalcore/_data_finder.py#L244
So something similar needs to happen for CORDEX, and possibly CMIP6. Trouble is I'm not sure the fix is so easy. I.e. some models have chosen to store fx variables under the same ensemble as the regular variables, whereas others have kept with the CMIP5 convention and left fx variables under r0i0p0.
I'll have a closer look through the datasets and see if I can summarise the situation.

**Comment by thomascrocker** [comment]
OK, I think I have developed a fix for this now. Just putting together a PR.
Tagging @valeriupredoi and @schlunma since it involves a small modification to https://github.com/ESMValGroup/ESMValCore/blob/22043381b6405c17c407a35bff8bf6ebc538b07d/esmvalcore/_data_finder.py#L242 which they appear to be authors of

**Comment by bouweandela** [comment]
> Are there suggestions for linting / code formatting for ESMValTool development? 

Please have a look at our [contribution guidelines](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/contributing.html#python).

**Comment by bouweandela** [comment]
Fixed by #1609.
